### PR TITLE
[8.x] Fix: Update Single JSON Field with Guarded

### LINKED ISSUE #37724: Update JSON Attribute blocked by Guarded Check
- Laravel Version: 8.47.0
- PHP Version: 7.4.14
- Database Driver & Version:

### Description:
The JSON `->` update helper, as described here: https://laravel.com/docs/8.x/eloquent-mutators#array-and-json-casting, does not work. 

```
$user = User::find(1);

$user->update(['options->key' => 'value']);
```

### Steps To Reproduce:
1. Create a model with `protected $guarded = ['id'];` and `protected $casts = [ 'options' => 'array'];`
2. Grab a model instance. `$user = $user->find(1);`
3. Attempt to update a nested value: `$user->update(['options->dinner' => 'pizza']);`

### Area of Code
This fails in the trait `Illuminate\Database\Eloquent\Concerns\GuardsAttributes` because when checking `$this->isFillable($key)` the `$this->isGuarded($key)` will return `true`. 

In this part of logic you are guarding anything that _is not_ a "valid, guardable column" or basically in the database schema. 
LINE 208: `! $this->isGuardableColumn($key = 'options->dinner')` 

But the issue is that a key of `'options->dinner'` will not be in the database. 

### Solution
Taylor removed this check for JSON: https://github.com/laravel/framework/commit/897d107775737a958dbd0b2f3ea37877c7526371

I think the solution is just to add it back:
```
/**
     * Determine if the given key is guarded.
     *
     * @param  string  $key
     * @return bool
     */
    public function isGuarded($key)
    {
        if (empty($this->getGuarded())) {
            return false;
        }
        
       // < -- here -->  
        if (strpos($key, '->') !== false) {
             $key = Str::before($key, '->');
        }

        return $this->getGuarded() == ['*'] ||
               ! empty(preg_grep('/^'.preg_quote($key).'$/i', $this->getGuarded())) ||
               ! $this->isGuardableColumn($key);
    }

```

### PR DESCRIPTION
**Pull request by jonrobinson**
Closes #37724 

This resolves a bug preventing the updating of a JSON field attribute when the a model `$guarded` has specified values. 
```
class User extends Model
{
    protected $guarded = ['id'];
    ...
```

The following is currently broken because `'options->key'` will not appear among the `isGuardableColumn()`
```
$user = User::find(1);

$user->update(['options->key' => 'value']);
```

### CODE REVIEW

**Comment by taylorotwell** [comment]
This introduces a security vulnerability. See previous PRs.

**Comment by jonrobinson** [comment]
@taylorotwell Thank you. I see that now. However, from my understanding, the vulnerability was that you could previously potentially mass assign on a guarded field. For instance, the issue was if you had `$guarded = ['options']` and passed `options->key` in a request and mass assigned it would not be blocked.

But you resolved this by adding the line `! $this->isGuardableColumn($key)` checking if it is a guardable column and rejecting anything not in the schema. 

Unless I'm missing something, adding this line keeps this protection while allowing you to use the -`>update([`options->key`])` on a non-guarded field (one you wouldn't use from the request). I confirmed this by modulating in my test, adding `arrayAttribute` to the guarded list, will prevent you from saving to the json field. 

In other words, when you add the following:
```
if (strpos($key, '->') !== false) {
    $key = Str::before($key, '->');
}
```
$key of `options->key` becomes `options`. If `options` is guarded it will be blocked. If `options` isn't guarded it will save. If you passed something like `options.key` it will be blocked (again by the ` $this->isGuardableColumn($key)` check).

I realize there's other ways of saving this, but I did like how clean the `$user->update(['options->key' => 'value']);` is.
